### PR TITLE
dicts should always be returned for TranslatedFields; add gate for old behavior

### LINKED ISSUE #8794: Always return (and only accept) a dict for localised fields
The current api/v3 behaviour of returning a flat string when `?lang=xx-YY` is passed means the api consumer doesn't know what language the string return actually is.  Accepting a flat string as input to update/create a localised field means the client doesn't really know what locale it's setting either (unless the `default_locale` is also set at the same time), and makes deleting localised values extra hard.

So we should drop that behaviour in v4.  Dicts will always be returned; dicts will always be required to set/update values.

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #8794 (but alas deletes are still as broken)

(Most of the change is tests and docs really)

### CODE REVIEW

**Comment by diox** [comment]
(submitted review by accident instead of just posting comment, not done yet with the review)

File: src/olympia/api/fields.py [hunk 1]
```diff
@@ -65,21 +78,27 @@ class TranslationSerializerField(fields.Field):
     default_error_messages = {
         'min_length': _(u'The field must have a length of at least {num} '
                         u'characters.'),
-        'unknown_locale': _(u'The language code {lang_code} is invalid.')
+        'unknown_locale': _(u'The language code {lang_code} is invalid.'),
+        'no_dict': _(u'You must provide a dictionary of {lang-code:value}.')
```
Comment (diox) [inline src/olympia/api/fields.py]: s/a dictionary/an object/ (json-speak)

File: docs/topics/api/overview.rst [hunk 2]
```diff
@@ -134,18 +134,41 @@ as keys and translations as values:
...
-This behaviour also applies to ``POST``, ``PATCH`` and ``PUT`` requests: you
-can either submit an object containing several translations, or just a string.
-If only a string is supplied, it will only be used to translate the field in
-the current language.
+Then the resulting translations would be::
+
+"name": {"en-US": "Fun", "fr": "Jeux","kn": "ಆಟಗಳು"}
+
+To delete a translation, pass ``null`` as the value for that language.
+(Note: this behavior is currently buggy/broken)
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply docs/topics/api/overview.rst]: Just filed https://github.com/mozilla/addons-server/issues/8816
Do you mean in the docs too?

[the remark this answers is not in the document]
Comment (diox) [reply docs/topics/api/overview.rst]: Yeah, IMHO we should reference the issue in the docs if we explicitly say in the docs it's buggy/broken.

[the remark this answers is not in the document]
Comment (eviljeff) [reply docs/topics/api/overview.rst]: https://github.com/mozilla/addons-server/pull/8795/commits/0405aeb5fc9f168fb5938fcd25cd591f723c4728

**Pull request merged**